File: src/lib/api/stream.ts
/**
 * Consumes the backend's server-sent analysis stream.
 *
 * Written by hand rather than with `EventSource` for one reason that matters:
 * `EventSource` cannot report *why* a connection failed. It fires a bare
 * `error` with no status and no body, so a backend that is simply not running
 * would be indistinguishable from one that rejected the address — and this
 * console's whole contract is that a failure says what went wrong and what to
 * do about it. `fetch` keeps the `ApiError` envelope every other call returns.
 */

import { ApiError, API_BASE_URL } from "@/lib/api/client";
import { authHeaders } from "@/lib/api/credentials";
import type {
  AnalysisTarget,
  ApiErrorBody,
  EvidenceItem,
  ExecutionResponse,
  SourceRef,
} from "@/lib/api/types";

/** Everything the run has established; only the summary is still pending. */
export type EvidenceReady = {
  target: AnalysisTarget | null;
  address: string | null;
  facts: Record<string, unknown>;
  evidence: EvidenceItem[];
  sources: SourceRef[];
};

export type StreamHandlers = {
  onStageStarted?: (stage: string) => void;
  onStageFinished?: (stage: string, durationMs: number | null) => void;
  onEvidenceReady?: (payload: EvidenceReady) => void;
  onSummaryDelta?: (text: string) => void;
};

type StreamRequest = {
  target: AnalysisTarget;
  /** A 0x address, a URL for `site`, or omitted entirely for `project`. */
  subject?: string;
  signal?: AbortSignal;
} & StreamHandlers;

/** SSE frames are separated by a blank line; fields are `name: value`. */
function parseFrame(frame: string): { event: string; data: string } | null {
  let event = "message";
  const data: string[] = [];

  for (const line of frame.split("\n")) {
    // A line starting with ':' is a comment — the keep-alive the server sends
    // so proxies do not close an idle connection mid-analysis.
    if (line.startsWith(":")) continue;
    if (line.startsWith("event:")) event = line.slice(6).trim();
    else if (line.startsWith("data:")) data.push(line.slice(5).trim());
  }

  return data.length ? { event, data: data.join("\n") } : null;
}

/**
 * Runs one analysis, reporting progress as it happens.
 *
 * Resolves with the complete result — identical to what the REST routes
 * return — so a caller that ignores every handler still gets the whole thing.
 */
export async function streamAnalysis({
  target,
  subject,
  signal,
  onStageStarted,
  onStageFinished,
  onEvidenceReady,
  onSummaryDelta,
}: StreamRequest): Promise<ExecutionResponse> {
  const params = new URLSearchParams({ target });
  if (subject) params.set("subject", subject);

  let response: Response;
  try {
    response = await fetch(`${API_BASE_URL}/api/v1/stream?${params.toString()}`, {
      // Analyses are metered, so the stream authenticates exactly like every
      // other call. This is also why `EventSource` is unusable here — it
      // cannot send an Authorization header at all.
      headers: { accept: "text/event-stream", ...authHeaders() },
      signal,
      cache: "no-store",
    });
  } catch (caught) {
    if (caught instanceof DOMException && caught.name === "AbortError") throw caught;
    throw new ApiError({
      code: "network_unreachable",
      message: "Cannot reach the Molthood API.",
      suggestedAction: `Start the backend with \`uvicorn app.main:app\` and confirm it is on ${API_BASE_URL}.`,
      status: 0,
    });
  }

  if (!response.ok || !response.body) {
    let payload: ApiErrorBody | null = null;
    try {
      payload = (await response.json()) as ApiErrorBody;
    } catch {
      // Not the standard envelope; fall through to a generic message.
    }
    throw new ApiError({
      code: payload?.error.code ?? "http_error",
      message: payload?.error.message ?? `The stream could not be opened (${response.status}).`,
      suggestedAction: payload?.error.suggested_action ?? "Retry the analysis.",
      status: response.status,
      requestId: payload?.request_id ?? null,
    });
  }

  const reader = response.body.pipeThrough(new TextDecoderStream()).getReader();
  let buffer = "";
  let result: ExecutionResponse | null = null;
  let failure: ApiError | null = null;

  try {
    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += value;

      // A chunk can split a frame anywhere, so only complete frames — those
      // terminated by a blank line — are dispatched. The remainder stays in
      // the buffer until the rest of it arrives.
      let boundary = buffer.indexOf("\n\n");
      while (boundary !== -1) {
        const frame = buffer.slice(0, boundary);
        buffer = buffer.slice(boundary + 2);
        boundary = buffer.indexOf("\n\n");

        const parsed = parseFrame(frame);
        if (!parsed) continue;

        let payload: Record<string, unknown>;
        try {
          payload = JSON.parse(parsed.data) as Record<string, unknown>;
        } catch {
          // One malformed frame must not abandon a run that is nearly done.
          continue;
        }

        switch (parsed.event) {
          case "stage_started":
            onStageStarted?.(String(payload.stage));
            break;
          case "stage_finished":
            onStageFinished?.(
              String(payload.stage),
              typeof payload.duration_ms === "number" ? payload.duration_ms : null,
            );
            break;
          case "evidence_ready":
            onEvidenceReady?.(payload as unknown as EvidenceReady);
            break;
          case "summary_delta":
            if (typeof payload.text === "string") onSummaryDelta?.(payload.text);
            break;
          case "result":
            result = payload as unknown as ExecutionResponse;
            break;
          case "error":
            failure = new ApiError({
              code: String(payload.code ?? "execution_failed"),
              message: String(payload.message ?? "The analysis failed."),
              suggestedAction: "Retry, or check the API logs.",
              status: 500,
            });
            break;
        }
      }
    }
  } finally {
    // Releases the connection when the caller aborts mid-run, so the backend
    // stops working for a reader who has navigated away.
    reader.cancel().catch(() => undefined);
  }

  if (failure) throw failure;
  if (!result) {
    throw new ApiError({
      code: "stream_incomplete",
      message: "The stream ended before the analysis returned a result.",
      suggestedAction: "Retry the analysis.",
      status: 0,
    });
  }

  return result;
}

File: backend/app/api/v1/endpoints/stream.py
"""`/api/v1/stream` — one analysis, delivered as it happens.

The same work as the plain `GET /token/{address}`, reported in stages instead
of withheld until all of it is done. That matters because the time is not
evenly spread: the evidence is complete in roughly a third of the run and then
waits behind an AI summary nobody asked to block on. Streaming does not make an
analysis faster — it stops the finished parts being held hostage by the slow
one.

Events, in the order a client sees them:

    stage_started    a stage began; `stage` names it
    stage_finished   that stage's outcome, with its duration
    evidence_ready   every finding, source and fact — the summary is all that
                     is left, so the report can be rendered here
    summary_delta    one fragment of generated prose
    result           the complete stored response, identical to the REST route
    error            the run failed; `message` says why

`result` always arrives, success or failure, so a client has exactly one thing
to wait for.
"""

from __future__ import annotations

import asyncio
import json
from collections.abc import AsyncIterator
from typing import Any

from fastapi import APIRouter, Query
from fastapi.responses import StreamingResponse

from app.api.auth import CurrentKey
from app.api.deps import EngineDep
from app.api.v1.endpoints.execute import to_response
from app.core.exceptions import MolthoodError, UnresolvableHostError
from app.engine.engine import ExecutionEngine
from app.engine.labels import redact_facts, redact_items
from app.engine.router import AnalysisTarget
from app.logging import get_logger
from app.repositories.api_keys import KeyIdentity, get_api_key_store
from app.services.web.fetcher import normalize_url, validate_public_url_async
from app.utils.validation import validate_address

logger = get_logger(__name__)

router = APIRouter(tags=["execution"])

#: Bounded so a client that stops reading cannot make the producer accumulate
#: an unbounded backlog of deltas in memory.
_QUEUE_SIZE = 256

#: Sent when nothing has happened for a while. Some proxies close an idle
#: connection, and a comment line resets that timer without being an event.
_HEARTBEAT_SECONDS = 15.0


def _sse(event: str, payload: dict[str, Any]) -> str:
    # `default=str` because facts carry datetimes, and one unserialisable value
    # must not abort a stream that is otherwise fine.
    body = json.dumps(payload, default=str)
    return f"event: {event}\ndata: {body}\n\n"


async def _resolve_address(target: AnalysisTarget, raw: str | None) -> str | None:
    """Validate the subject exactly as the non-streaming routes do."""
    if target is AnalysisTarget.PROJECT:
        return None

    if raw is None or not raw.strip():
        raise MolthoodError(
            f"A {target.value} analysis needs a subject.",
            details={"target": target.value},
        )

    if target is AnalysisTarget.SITE:
        # A private or malformed host is a caller error; one that simply does
        # not resolve is a finding, and the agent reports it.
        try:
            return await validate_public_url_async(raw)
        except UnresolvableHostError:
            return normalize_url(raw)

    return validate_address(raw)


async def _events(
    engine: ExecutionEngine,
    target: AnalysisTarget,
    address: str | None,
    identity: KeyIdentity,
) -> AsyncIterator[str]:
    queue: asyncio.Queue[tuple[str, dict[str, Any]] | None] = asyncio.Queue(_QUEUE_SIZE)

    async def emit(event: str, payload: dict[str, Any]) -> None:
        # `result` is redacted by `to_response` on the way out, but nothing was
        # doing the same for `evidence_ready` — and that one arrives first and
        # carries the findings the console renders immediately. Supplier names
        # were visible for the whole length of an AI summary and then silently
        # replaced when the run finished.
        if event == "evidence_ready":
            payload = {
                **payload,
                "facts": redact_facts(payload.get("facts")),
                "evidence": redact_items(payload.get("evidence") or []),
                "sources": redact_items(payload.get("sources") or []),
            }
        await queue.put((event, payload))

    async def produce() -> None:
        store = get_api_key_store()
        charged = identity.id != "open-mode"

        try:
            if charged:
                await store.consume(identity.id)
        except MolthoodError as exc:
            # Quota is checked inside the stream rather than in a dependency so
            # the refusal arrives as an `error` event the console can render,
            # not as a failed connection with no explanation.
            await queue.put(
                ("error", {"code": exc.code, "message": exc.message, **exc.details})
            )
            await queue.put(None)
            return

        try:
            result = await engine.analyze(
                target=target.value,
                address=address,
                emitter=emit,
                owner_key_id=identity.id,
            )
            await queue.put(("result", to_response(result).model_dump(mode="json")))
        except MolthoodError as exc:
            if charged:
                await store.refund(identity.id)
            await queue.put(("error", {"code": exc.code, "message": exc.message}))
        except asyncio.CancelledError:
            # The reader left before the run finished. It never produced a
            # result they could use, so the unit goes back.
            if charged:
                await store.refund(identity.id)
            raise
        except Exception as exc:
            if charged:
                await store.refund(identity.id)
            logger.exception("stream_execution_failed", error=str(exc))
            await queue.put(("error", {"code": "internal", "message": str(exc)}))
        finally:
            await queue.put(None)

    task = asyncio.create_task(produce())

    try:
        while True:
            try:
                item = await asyncio.wait_for(queue.get(), timeout=_HEARTBEAT_SECONDS)
            except TimeoutError:
                yield ": keep-alive\n\n"
                continue

            if item is None:
                return

            event, payload = item
            yield _sse(event, payload)
    finally:
        # Reached when the client disconnects mid-run. Without this the
        # execution would keep working — and keep spending credit — for a
        # reader who has already closed the tab.
        if not task.done():
            task.cancel()
        await asyncio.gather(task, return_exceptions=True)


@router.get(
    "/stream",
    summary="Run an analysis and stream its progress",
    response_class=StreamingResponse,
    description=(
        "Server-sent events for one analysis. Emits `stage_started`, "
        "`stage_finished`, `evidence_ready`, `summary_delta`, and finally "
        "`result` — the same object the REST routes return."
    ),
)
async def stream_analysis(
    engine: EngineDep,
    identity: CurrentKey,
    target: AnalysisTarget = Query(description="What kind of subject to analyse."),
    subject: str | None = Query(
        default=None,
        description="A 0x address, or a URL when target is `site`. Omit for `project`.",
    ),
) -> StreamingResponse:
    address = await _resolve_address(target, subject)

    return StreamingResponse(
        _events(engine, target, address, identity),
        media_type="text/event-stream",
        headers={
            "cache-control": "no-cache, no-transform",
            "connection": "keep-alive",
            # nginx buffers proxied responses by default, which would hold every
            # event until the run finished and defeat the entire endpoint.
            "x-accel-buffering": "no",
        },
    )
